### PR TITLE
Improve text coloring across `normalize` modes for confusion matrix (≤30 classes)

### LINKED ISSUE #20948: Hard coded black/white cutoff leads to low contrast non-normalized confusion matrix
The issue is caused by this line:
https://github.com/ultralytics/ultralytics/blob/9f4eb2491b6a25abaae29b74da6698cefe8dc662/ultralytics/utils/metrics.py#L478

This example can reproduce the issue:
```pthon
from ultralytics.utils.metrics import ConfusionMatrix
import numpy as np

conf = ConfusionMatrix(2)
conf.matrix = np.array([[450, 0], [0, 20]])
conf.plot(normalize=False, save_dir="conftest")
```

Image

Possible fix, make the value for the non-normalized case dependent on the maximum value in the array:
`color="white" if val > (0.7 if normalize else np.nanmax(array)*0.7) else "black",`

Image

### PR DESCRIPTION
**Pull request by mihlefeld**
I have read the CLA Document and I sign the CLA

This pull request fixes #20948 and improves contrast of non-normalized confusion matrices.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Improved the readability of confusion matrix plots by adjusting text color based on cell values.

### 📊 Key Changes  
- Updated the logic for choosing text color in confusion matrix cells, using a dynamic threshold for better contrast.
- The threshold now adapts based on whether the matrix is normalized, making white text appear only on higher-value cells.

### 🎯 Purpose & Impact  
- 🖼️ Enhances the visual clarity of confusion matrix plots, making them easier to interpret.
- 👀 Users can more easily read cell values, especially in cases with varying data scales.
- 🚀 Improves overall user experience when analyzing model performance.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mihlefeld, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and follow the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If your PR addresses a bug and you have not already supplied one, please provide a minimum reproducible example (MRE) to help us verify the fix. Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🌟

_This is an automated response. An Ultralytics engineer will review your PR and assist you soon!_

**Comment by glenn-jocher** [comment]
@RizwanMunawar please review

**Comment by RizwanMunawar** [comment]
Hi @mihlefeld, thanks for the PR! I reviewed the example you shared regarding the issue. Could you please provide a specific use case along with the dataset you're using, such as an object detection dataset? I'm unable to reproduce the results with the coco8 dataset.

```python
from ultralytics import YOLO

# Load a pretrained YOLO11n model
model = YOLO("yolo11n.pt")

# Train the model on COCO8
results = model.train(data="coco8.yaml", epochs=3, imgsz=640)
```

I also think we don't need to use np.nanmax(array) `ax.imshow` automatically set vmax based on the array's max value. We can use that directly. Also, can you please sign the CLA? Thanks

**Review by RizwanMunawar — COMMENTED** [review]
Hi @mihlefeld, thanks for the PR! I reviewed the example you shared with the issue. Could you please provide a specific use case along with the dataset you're using, such as an object detection dataset? I'm unable to reproduce the results with the coco8 dataset.

```python
from ultralytics import YOLO

# Load a pretrained YOLO11n model
model = YOLO("yolo11n.pt")

# Train the model on COCO8
results = model.train(data="coco8.yaml", epochs=3, imgsz=640)
```

I also think we don't need to use np.nanmax(array) `ax.imshow` automatically set vmax based on the array's max value. We can use that directly. Also, can you please sign the CLA? Thanks

**Comment by mihlefeld** [comment]
I have read the CLA Document and I sign the CLA

**Comment by mihlefeld** [comment]
recheck

**Comment by mihlefeld** [comment]
> Hi @mihlefeld, thanks for the PR! I reviewed the example you shared with the issue. Could you please provide a specific use case along with the dataset you're using, such as an object detection dataset? I'm unable to reproduce the results with the coco8 dataset.
> 
> ```python
> from ultralytics import YOLO
> 
> # Load a pretrained YOLO11n model
> model = YOLO("yolo11n.pt")
> 
> # Train the model on COCO8
> results = model.train(data="coco8.yaml", epochs=3, imgsz=640)
> ```
> 
> I also think we don't need to use np.nanmax(array) `ax.imshow` automatically set vmax based on the array's max value. We can use that directly. Also, can you please sign the CLA? Thanks

```python
from ultralytics import YOLO

# Load pretrained model
model = YOLO("yolo11n.pt")

# Train the model on HomeObjects-3K dataset
model.train(data="HomeObjects-3K.yaml", epochs=2, imgsz=640, project="runs")
```
image

New version:
image

**Comment by mihlefeld** [comment]
Nanmax is necessary because of 
https://github.com/mihlefeld/ultralytics/blob/eae140350e09c572679864c36e72eb2cfcf83026/ultralytics/utils/metrics.py#L443 
`array[array < 0.005] = np.nan  # don't annotate (would appear as 0.00)`
And `np.array([np.nan, 1]).max()` results in `np.nan` which then makes black the default font color, so you lose contrast in the dark blues:
image

**Comment by RizwanMunawar** [comment]
Ok, got it. Thanks for sharing. I think in the example you shared, alot of black text I see, which ideally we don't want. Yes agree, few places we need to adjust. I think the `vmax` value should work. Let me test, and I will update the PR.

**Review by RizwanMunawar — APPROVED** [review]

**Comment by RizwanMunawar** [comment]
@glenn-jocher @mihlefeld this PR is good to merge.

- @glenn-jocher I double checked our previous implementation with seaborn and noticed that, `0.45` threshold has been used there for coloring, adjusted this PR accordingly for `normalize=True`. Additionally, we are using the hardcoded value `2` (cmap vmax value), in case of non-normalized confusion matrix, @mihlefeld replaced it with `np.nanmax(array)`, and it will align both metrics with the same colour adjustment.

I attached the side-by-side results on the implementation we previously had with **Seaborn** vs the **Existing implementation** vs the change with **This PR**. Thanks

 
### Normalized=False

| Seaborn Output | Existing Implementation | This PR |
|----------------|--------------------------|---------|
| Seaborn | Existing | This PR |



### Normalized=True
| Seaborn Output (Normalized) | Existing Implementation (Normalized) | This PR (Normalized) |
|-----------------------------|---------------------------------------|------------------------|
| Seaborn | Existing | This PR |

**Comment by UltralyticsAssistant** [comment]
Exciting news—your PR is merged! 🚀 As Leonardo da Vinci said, “Simplicity is the ultimate sophistication.” Your thoughtful improvements to the confusion matrix plots embody this wisdom, making complex results clearer and more accessible for all Ultralytics users. Thank you @mihlefeld, with valuable contributions from @glenn-jocher and @RizwanMunawar, for enhancing the user experience and raising the bar for clarity in model analysis. Your work truly makes a difference!
